# IPython/Jupyter magic function documentation - `%watermark`

I wrote this simple `watermark` IPython magic function to conveniently add date- and time-stamps to my IPython notebooks. Also, I often want to document various system information, e.g., for my [Python benchmarks](https://github.com/rasbt/One-Python-benchmark-per-day) series.


&nbsp;
## Installation

The watermark line magic can be installed by executing

```bash
pip install watermark
```

Alternatively, you can install the latest development version directly from GitHub via

```bash
pip install -e git+https://github.com/rasbt/watermark#egg=watermark
```

&nbsp;
## Loading the `%watermark` magic

To load the `watermark` magic, execute the following line in your IPython notebook or current IPython shell

In [1]:
%load_ext watermark

The version of `watermark` itself can be retrieved as follows:

In [2]:
%watermark --watermark

Watermark: 2.5.1



&nbsp;
## Usage

In order to display the optional `watermark` arguments, type

In [3]:
%watermark?

Docstring:
::

  %watermark [-a AUTHOR] [-gu GITHUB_USERNAME] [-e EMAIL] [-ws WEBSITE]
                 [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME] [-v]
                 [-p PACKAGES] [-co] [-h] [-m] [-g] [-r] [-b] [-w] [-iv]
                 [--gpu]

IPython magic function to print date/time stamps
and various system information.

options:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -gu GITHUB_USERNAME, --github_username GITHUB_USERNAME
                        prints author github username
  -e EMAIL, --email EMAIL
                        prints author email
  -ws WEBSITE, --website WEBSITE
                        prints author or project website
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone in the ISO 8601 standar

<pre>
%watermark [-a AUTHOR] [-gu GITHUB_USERNAME] [-e EMAIL] [-ws WEBSITE]
                 [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME] [-v]
                 [-p PACKAGES] [-h] [-m] [-g] [-r] [-b] [-w] [-iv]

IPython magic function to print date/time stamps
and various system information.

options:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -gu GITHUB_USERNAME, --github_username GITHUB_USERNAME
                        prints author github username
  -e EMAIL, --email EMAIL
                        prints author email
  -ws WEBSITE, --website WEBSITE
                        prints author or project website
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone in the ISO 8601 standard with UTC offset
  -z, --timezone        appends the local time zone
  -u, --updated         appends a string "Last updated: "
  -c CUSTOM_TIME, --custom_time CUSTOM_TIME
                        prints a valid strftime() string
  -v, --python          prints Python and IPython version
  -p PACKAGES, --packages PACKAGES
                        prints versions of specified Python modules and
                        packages
  -h, --hostname        prints the host name
  -m, --machine         prints system and machine info
  -g, --githash         prints current Git commit hash
  -r, --gitrepo         prints current Git remote address
  -b, --gitbranch       prints current Git branch
  -w, --watermark       prints the current version of watermark
  -iv, --iversions      prints the name/version of all imported modules
</pre>

&nbsp;
## Examples

&nbsp;
### Defaults

In [4]:
%watermark

Last updated: 2025-12-14T11:40:23.920613-06:00

Python implementation: CPython
Python version       : 3.13.5
IPython version      : 9.8.0

Compiler    : Clang 20.1.4 
OS          : Darwin
Release     : 24.6.0
Machine     : arm64
Processor   : arm
CPU cores   : 12
Architecture: 64bit



&nbsp;
### Last updated date and time

In [5]:
%watermark -u -t -d

Last updated: 2025-12-14 11:40:23



In [6]:
%watermark -u --iso8601

Last updated: 2025-12-14T11:40:23.932024-06:00



In [7]:
%watermark -u -n -t -z

Last updated: Sun Dec 14 2025 11:40:23CST



&nbsp;
### Python version information

In [8]:
%watermark -v

Python implementation: CPython
Python version       : 3.13.5
IPython version      : 9.8.0



&nbsp;
### Machine information

In [9]:
%watermark -m

Compiler    : Clang 20.1.4 
OS          : Darwin
Release     : 24.6.0
Machine     : arm64
Processor   : arm
CPU cores   : 12
Architecture: 64bit



&nbsp;
### Combination of information

In [10]:
%watermark -v -m -p numpy,scipy,sklearn 

Python implementation: CPython
Python version       : 3.13.5
IPython version      : 9.8.0

numpy  : 2.3.5
scipy  : 1.16.3
sklearn: 1.8.0

Compiler    : Clang 20.1.4 
OS          : Darwin
Release     : 24.6.0
Machine     : arm64
Processor   : arm
CPU cores   : 12
Architecture: 64bit



&nbsp;
### With author

In [11]:
%watermark -a "John Doe" -u -d -v -m

Author: John Doe

Last updated: 2025-12-14

Python implementation: CPython
Python version       : 3.13.5
IPython version      : 9.8.0

Compiler    : Clang 20.1.4 
OS          : Darwin
Release     : 24.6.0
Machine     : arm64
Processor   : arm
CPU cores   : 12
Architecture: 64bit



&nbsp;
### Imported modules information

In [12]:
import scipy as sp
from sklearn import metrics
import numpy.linalg as linalg

In [13]:
%watermark --iversions

numpy  : 2.3.5
scipy  : 1.16.3
sklearn: 1.8.0

